### PR TITLE
Add embedding and MLP support for sparse wide features

### PR DESCRIPTION
**Pull request by StarWang**
# Description

Currently DeText's design for sparse feature has simple modeling power for sparse features.
1. only linear model is applied on sparse features
2. there's no interaction between sparse features and dense features (model_score = dense_score + sparse_score)

This PR resolves the above limitation on sparse feature by
1. computing dense representation of sparse features
2. allowing interactions between sparse features and wide features

More specifically, the model architecture changes from
```
dense_score = dense_ftrs -> MLP
sparse_score = sparse_ftrs -> Linear
final_score = dense_score + sparse_score
```
to
```
sparse_emb_ftrs = sparse_ftrs -> Dense(sp_emb_size)
all_ftrs = (dense_ftrs, sparse_emb_ftrs) -> Concatenate
final_score= all_ftrs -> MLP
```
## Type of change

- [ ] New feature (non-breaking change which adds functionality)

## List all changes 
Please list all changes in the commit.
* Change sp_linear_model to sp_emb_model and add an option sp_emb_size to allow the sparse matrix to have output dimension > 1
* Change structure of dense & sparse feature interaction as mentioned in the PR description
* Add and restructure unit test for sparse embedding model
* Add new data for testing
* Add a sample tfrecord generation helper function in misc_utils.py

# Testing
- Successfully run run_detext.sh for data including wide_sp_val and sp_emb_size=10
- Successfully run run_detext_multitask.sh for data
- Unit test for sparse_emb_model when sp_emb_size is 1 and > 1
# Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by zhoutong-fu** [comment]
Two general comments: 1) in either the description or readme file update the latest model structure (can use the new feature Xiaowei added last time that dumps the structure to a txt file; 2) can you check if run_ detext_multitask.sh works?

File: src/detext/utils/misc_utils.py [hunk 1]
```diff
@@ -406,84 +406,52 @@ def estimate_train_steps(input_pattern, num_epochs, batch_size, isTfrecord):
...
+    """Returns a float_list feature"""
+    return tf.train.Feature(float_list=tf.train.FloatList(value=value))
+
+
+def _int64_feature(value):
+    """Returns an int64_list feature"""
+    return tf.train.Feature(int64_list=tf.train.Int64List(value=value))
+
+
+def create_sample_tfrecord(out_file):
```
Comment (zhoutong-fu) [inline src/detext/utils/misc_utils.py:426]: If these functions are utils for testing, shall we move it to `test_utils.py`? It'll be good to clean up some functions in `test_utils.py` if needed.

File: src/detext/utils/misc_utils.py [hunk 2]
```diff
@@ -406,84 +406,52 @@ def estimate_train_steps(input_pattern, num_epochs, batch_size, isTfrecord):
...
-    #     '/home/wguo/data/people_v6/NAV/train/00005',
-    #     '/home/wguo/data/people_v6/NAV/train/00006',
-    #     '/home/wguo/data/people_v6/NAV/train/00007',
-    # ]
-    # output_file = '/home/wguo/data/people_v6/NAV/mean_std.pkl'
-    # get_mean_std(input_files, output_file, 167)
-    # # shuffle_tfrecords(input_file, output_file)
-    # # random_baseline('/home/xwli/ranking/job-search-data/dev/gen-tfrecords/output1.tfrecords', 25)
-    #
-    # # for people search latency test
```
Comment (zhoutong-fu) [inline src/detext/utils/misc_utils.py:433]: it's great to clean up the unused code or comments 👍

File: src/detext/model/sp_emb_model.py [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 
     def __init__(self,
                  num_wide_sp: int,
                  wide_ftrs_sp_idx: tf.Tensor,
+                 sp_emb_size: int,
                  wide_ftrs_sp_val: tf.Tensor = None,
+                 padding_idx: int = 0,
                  initializer=tf.contrib.layers.xavier_initializer()):
-        """ Computes a linear score given wide feature indices and values
+        """ Computes a embedding given wide feature indices and values
```
Comment (zhoutong-fu) [inline src/detext/model/sp_emb_model.py:14]: suggests adding sparse input features related instruction to the training data format section in `TRAINING.md`

File: src/detext/model/sp_emb_model.py [hunk 4]
```diff
@@ -26,25 +28,27 @@ def __init__(self,
...
         with tf.variable_scope('wide', reuse=tf.AUTO_REUSE):
             # Feature weights
             self.ftrs_weight = tf.get_variable('wide_ftrs_sp_weight',
-                                               shape=[num_wide_sp, 1],
+                                               shape=[num_wide_sp, sp_emb_size],
                                                initializer=initializer,
                                                trainable=True)
 
             # A hack to combine idx and val so that we can process them together in `tf.map_fn` later
+            # Shape=[batch_size, max_group_size, max_wide_ftrs_size*2]
```
Comment (zhoutong-fu) [inline src/detext/model/sp_emb_model.py:41]: Can you comment on what the `max_wide_ftrs_size` is? I think I saw related comments somewhere else but it's great to consolidate them under this class.

File: src/detext/utils/misc_utils.py [hunk 5]
```diff
@@ -406,84 +406,52 @@ def estimate_train_steps(input_pattern, num_epochs, batch_size, isTfrecord):
...
+    print("Composing fake tfrecord to file {}".format(out_file))
+    with tf.io.TFRecordWriter(out_file) as writer:
+        with tf.Graph().as_default(), tf.compat.v1.Session():
+            # Example 1
+            features = {
+                "label": _float_feature(
+                    [1., 0., 0., 0.],
+                ),
+                "query": _bytes_feature([b"hello"]),
+                "wide_ftrs": _float_feature([23.0, 14.0, 44.0, -1.0, 22.0, 19.0, 22.0, 19.0]),
```
Comment (xwli-chelsea) [inline src/detext/utils/misc_utils.py]: This is useful 👍 . Can we also add a user field?

File: src/detext/utils/misc_utils.py (hunk 1 above)
Comment (StarWang) [reply src/detext/utils/misc_utils.py]: Moved to test_utils.py in latest commit

File: src/detext/model/sp_emb_model.py (hunk 3 above)
Comment (StarWang) [reply src/detext/model/sp_emb_model.py:14]: Added in TRAINING.md in latest update

File: src/detext/model/sp_emb_model.py [hunk 6]
```diff
@@ -1,15 +1,17 @@
...
-    """Linear model that performs computation on sparse features"""
+class SparseEmbModel:
+    """Embedding model that performs embedding lookup and summation on sparse features"""
 
     def __init__(self,
                  num_wide_sp: int,
                  wide_ftrs_sp_idx: tf.Tensor,
+                 sp_emb_size: int,
                  wide_ftrs_sp_val: tf.Tensor = None,
+                 padding_idx: int = 0,
```
Comment (xwli-chelsea) [inline src/detext/model/sp_emb_model.py:12]: Add comment for the params

File: src/detext/model/sp_emb_model.py (hunk 4 above)
Comment (StarWang) [reply src/detext/model/sp_emb_model.py]: Added comment in this line

**Comment by StarWang** [comment]
> Two general comments: 1) in either the description or readme file update the latest model structure (can use the new feature Xiaowei added last time that dumps the structure to a txt file; 2) can you check if run_ detext_multitask.sh works on multitask test data (not the sparse wide features test data)?

@zhoutong-fu Thanks for the review!
I've updated the readme file and tested that run_detext_multitask.sh runs successfully

File: src/detext/utils/misc_utils.py (hunk 5 above)
Comment (StarWang) [reply src/detext/utils/misc_utils.py]: Sure. Added in latest commit

File: src/detext/model/sp_emb_model.py (hunk 6 above)
Comment (StarWang) [reply src/detext/model/sp_emb_model.py:12]: Added in latest commit

File: TRAINING.md [hunk 7]
```diff
@@ -14,8 +14,34 @@ We show an example of the prepared training data and explain the data format and
...
-    * [0.305 0.264 0.180 0.192 0.136 0.027 0.273 0.273 0.377 0.233 0.264 0.227 0.119 0.119 0.198 0.212 0.274 0.047 0.320 0.255 0.350 0.000 0.000 0.357 0.301 0.367 0.292 0.000 0.000 0.170]
+    * There could be multiple dense wide features for each document. Therefore the dense wide features are a 2-D array 
+    with shape [#documents, #dense wide features per document]. Since TFRecords support 1-D FloatList, we flatten the 
+    dense wide features in the preparation and transform to grouped features in `train/data_fn.py` by reshaping. 
+    Therefore the float list of wide_ftrs in the training data has `#documents * #dense wide features per document = 4 * 3 = 12` entries. 
+    The dense wide features belong to each document sequentially. I.e., the first 3 wide features belong to the first 
+    document, the second 3 wide features belong to the second document, etc..
+    * [0.305 0.264 0.180 0.192 0.136 0.027 0.273 0.273 0.377 0.233 0.264 0.227]
+* `wide_ftrs_sp_idx` (int list)
```
Comment (zhoutong-fu) [inline TRAINING.md:27]: This is very helpful. Thanks for the documentation update.

**Pull request merged**